wallet/src/de/schildbach/wallet/ui/DecryptSeedWithPinDialog.kt

```diff
@@ -0,0 +1,114 @@
+            } else {
+                val args = Bundle()
+                args.putInt(ARG_REQUEST_CODE, requestCode)
+                args.putBoolean(ARG_PIN_ONLY, pinOnly)
+                checkPinDialog.arguments = args
+                checkPinDialog.show(activity.supportFragmentManager, FRAGMENT_TAG)
+            }
+        }
+
+        @JvmStatic
+        fun show(activity: AppCompatActivity, requestCode: Int = 0) {
+            show(activity, requestCode, false)
+        }
+
+    }
+
+    override fun initViewModel() {
+        viewModel = ViewModelProviders.of(this).get(DecryptSeedViewModel::class.java)
+        (viewModel as DecryptSeedViewModel).decryptSeedLiveData.observe(viewLifecycleOwner, Observer {
+            when (it.status) {
+                Status.ERROR -> {
+                    pinRetryController.failedAttempt(it.data!!.second!!)
+                    if (pinRetryController.isLocked) {
+                        showLockedAlert(context!!)
+                        dismiss()
+                        return@Observer
+                    }
+                    setState(State.INVALID_PIN)
+                }
+                Status.LOADING -> {
+                    setState(State.DECRYPTING)
+                }
+                Status.SUCCESS -> {
+                    dismiss(it.data!!.first!!, it.data!!.second!!)
+                }
+            }
+        })
+    }
+
+    private fun dismiss(seed : DeterministicSeed, pin: String) {
+        if (pinRetryController.isLocked) {
+            return
+        }
+        val requestCode = arguments!!.getInt(ARG_REQUEST_CODE)
+        (sharedModel as DecryptSeedSharedModel).onDecryptSeedCallback.value = Pair(requestCode, seed)
+        pinRetryController.clearPinFailPrefs()
+        dismiss()
+    }
+
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
+        super.onActivityCreated(savedInstanceState)
+        activity?.run {
+            sharedModel = ViewModelProviders.of(this)[DecryptSeedSharedModel::class.java]
+        } ?: throw IllegalStateException("Invalid Activity")
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    override fun startFingerprintListener() {
+        fingerprintCancellationSignal = CancellationSignal()
+        fingerprintHelper!!.getPassword(fingerprintCancellationSignal, object : FingerprintHelper.Callback {
+            override fun onSuccess(savedPass: String) {
+                //dismiss(savedPass)
+                (viewModel as DecryptSeedViewModel).checkPin(savedPass)
+            }
+
+            override fun onFailure(message: String, canceled: Boolean, exceededMaxAttempts: Boolean) {
+                if (!canceled) {
+                    fingerprint_view.showError(exceededMaxAttempts)
+                }
+            }
+
+            override fun onHelp(helpCode: Int, helpString: String) {
+                fingerprint_view.showError(false)
+            }
+        })
+    }
```

Comment:
@HashEngineering ,
in order to condense the code even more and avoid duplicates, instead of overriding the whole `startFingerprintListener()` method, you can "override" just the `onSuccess(savedPass: String)` by extracting its content into a separate method.
To be more clear. In the `CheckPinDialog` add the method like `onSuccessFingerprint(savedPass: String)` and call it instead of `dismiss(savedPass)`. The base version of this method (in CheckPinDialog) will then call `dismiss(savedPass)` whereas the overridden version (in `DecryptSeedWithPinDialog`) `(viewModel as DecryptSeedViewModel).checkPin(savedPass)`

Reply:
Excellent idea.

Reply:
I think we can do this in a few more places also.

Reply:
https://github.com/dashevo/dash-wallet/pull/300/commits/be532926bb667e8d4b425ba9f5b1375c7f808232